src/persistent-connections/content.md

```diff
@@ -0,0 +1,465 @@
+  "slides": [
+    {
+      "image": "assets/persistent-iptables-1.svg",
+      "description": "The red Pod issues a request to the Service."
+    },
+    {
+      "image": "assets/persistent-iptables-2.svg",
+      "description": "You already know what happens next. Services don't exist, but iptables rules intercept the requests."
+    },
+    {
+      "image": "assets/persistent-iptables-3.svg",
+      "description": "One of the Pods that belong the Service is selected as the destination."
+    },
+    {
+      "image": "assets/persistent-iptables-4.svg",
+      "description": "Finally the request reaches the Pod. At this point, a persistent connection between the two Pods is established."
+    },
+    {
+      "image": "assets/persistent-iptables-5.svg",
+      "description": "Any subsequet request from the red Pod reuses the existing open connection."
+    }
+  ]
+}
+```
+
+So you achieved better latency and throughput, but you lost the ability to scale your backend.
+
+Even if you have two backend Pods that can receive requests, only one is actively used.
+
+_Is it fixable?_
+
+Since Kubernetes doesn't know how to load balance persistent connections, you could step in and fix it yourself.
+
+Services are a collection of IP addresses and ports — usually called endpoints.
+
+Your app could retrieve the list of endpoints from the Service and decide how to distribute the requests.
+
+As a first try, you could open a persistent connection to every Pod and round-robin requests to them.
+
+Or you could [implement more sophisticated load balancing algorithms](https://blog.twitter.com/engineering/en_us/topics/infrastructure/2019/daperture-load-balancer.html).
+
+The client-side code that executes the load balancing should follow the following logic:
+
+1. retrieve a list of endpoints from the Service
+1. for each of them, open a connection and keep it open
+1. when you need to make a request, pick one of the open connections
+1. on a regular interval refresh the list of endpoints and remove or add new connections
+
+```slideshow
+{
+  "description": "Client side load balancing",
+  "slides": [
+    {
+      "image": "assets/client-lb-1.svg",
+      "description": "Instead of having the red Pod issuing a request to your Service, you could load balance the request client-side."
+    },
+    {
+      "image": "assets/client-lb-2.svg",
+      "description": "Your could write some code that asks what Pods are part of the Service."
+    },
+    {
+      "image": "assets/client-lb-3.svg",
+      "description": "Once you have that list, you could store it locally and use it to connect to the Pods."
+    },
+    {
+      "image": "assets/client-lb-4.svg",
+      "description": "You are in charge of the load balancing algorithm."
+    }
+  ]
+}
+```
+
+_Does this problem apply only to keep-alive?_
+
+## Database persistent connections
+
+HTTP isn't the only protocol designed to open a long-lived connection.
```

Comment:
HTTP isn't the only protocol that can benefit from long-lived TCP connections.

Reply:
I like this.